### PR TITLE
feat: slicing PyVector&Create DataFrame from sql

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

You can now slice PyVector in Pythonic ways, i.e:
`a[1:2]` `a[-2:-1]` `a[-1:-2:-1]`

And be able to create DataFrame from sql! This use `query().sql()` to get a `RecordBatch` and convert it to `Dataframe` :
```python
dataframe.from_sql(...)
```

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/script/src/python/pyo3/dataframe_impl.rs [hunk 1]
```diff
@@ -49,6 +50,15 @@ impl PyDataFrame {
...
 impl PyDataFrame {
+    #[classmethod]
+    fn from_sql(_cls: &PyType, py: Python, sql: String) -> PyResult<Self> {
+        let query = query_engine(py)?;
+        let rb = query.sql_to_rb(sql).map_err(PyRuntimeError::new_err)?;
+        let ctx = datafusion::execution::context::SessionContext::new();
+        ctx.read_batch(rb.df_record_batch().clone())
+            .map_err(|e| PyRuntimeError::new_err(format!("{e:?}")))
+            .map(Self::from)
+    }
```
Comment (waynexia) [inline src/script/src/python/pyo3/dataframe_impl.rs:61]: Great work!

File: src/script/src/python/ffi_types/pair_tests/sample_testcases.rs [hunk 2]
```diff
@@ -56,6 +57,34 @@ def boolean_array() -> vector[f64]:
...
+            expect: Some(ronish!("value": vector!(Float64Vector, [2.0f64]))),
+        },
+        CoprTestCase {
+            script: r#"
+@copr(returns=["value"], backend="rspy")
+def boolean_array() -> vector[f64]:
+    from greptime import vector
+    from greptime import query, dataframe, PyDataFrame
+    
+    df = PyDataFrame.from_sql("select number from numbers limit 5")
```
Comment (waynexia) [inline src/script/src/python/ffi_types/pair_tests/sample_testcases.rs:83]: can we add a case that combine SQL and dataframe API? like 
```python
df = PyDataFrame.from_sql("blabla").filter("blabla")
```
It should be supported now.

File: src/script/src/python/pyo3/copr_impl.rs [hunk 3]
```diff
@@ -165,6 +164,23 @@ coprocessor = copr
 /// Cast return of py script result to `Vec<VectorRef>`,
 /// constants will be broadcast to length of `col_len`
 fn py_any_to_vec(obj: &PyAny, col_len: usize) -> PyResult<Vec<VectorRef>> {
+    // 1. check if obj is of two types:
```
Comment (waynexia) [inline src/script/src/python/pyo3/copr_impl.rs:167]: nit: looks like there is no "2."?
```suggestion
    // check if obj is of two types:
```

File: src/script/src/python/pyo3/vector_impl.rs [hunk 4]
```diff
@@ -299,6 +299,51 @@ impl PyVector {
...
+            let mut buf = vector
+                .data_type()
+                .create_mutable_vector(indices.slicelength as usize);
+            let v = if indices.slicelength == 0 {
+                buf.to_vector()
+            } else if step == 1 {
+                vector.slice(start as usize, slicelength as usize)
+            } else {
+                if indices.step > 0 {
+                    let range = if stop == -1 {
```
Comment (waynexia) [inline src/script/src/python/pyo3/vector_impl.rs:326]: can this handle `v[:-2]`?

BTW, for simplicy, I think we need not special handle `1` (like step = 1 or stop = -1), and leave it to rust `Range`. We only need to take care of negative ranges. WDYT

File: src/script/src/python/ffi_types/copr.rs [hunk 5]
```diff
@@ -343,20 +343,43 @@ pub(crate) enum Either {
...
+                let rb = compute::concat_batches(
+                    rbs.schema().arrow_schema(),
+                    rbs.iter().map(|r| r.df_record_batch()),
+                )
+                .map_err(|e| {
+                    format!(
+                        "Concat batches failed for query {sql}: {e}",
+                        sql = sql,
+                        e = e
+                    )
```
Comment (MichaelScofield) [inline src/script/src/python/ffi_types/copr.rs:361]:
```suggestion
                    format!("Concat batches failed for query {sql}: {e}")
```

File: src/script/src/python/ffi_types/copr.rs [hunk 6]
```diff
@@ -343,20 +343,43 @@ pub(crate) enum Either {
...
+                })?;
+                RecordBatch::try_from_df_record_batch(rbs.schema(), rb).map_err(|e|
+                    format!(
+                        "Convert datafusion record batch to record batch failed for query {sql}: {e}",
+                        sql = sql,
+                        e = e
+                    )
```
Comment (MichaelScofield) [inline src/script/src/python/ffi_types/copr.rs:368]:
```suggestion
                    format!("Convert datafusion record batch to record batch failed for query {sql}: {e}")
```

File: src/script/src/python/pyo3/copr_impl.rs [hunk 7]
```diff
@@ -165,6 +164,23 @@ coprocessor = copr
...
+    // tuples of PyVector
+    // a single PyVector
+    let check = if obj.is_instance_of::<PyTuple>()? {
+        let tuple = obj.downcast::<PyTuple>()?;
+
+        (0..tuple.len())
+            .map(|idx| tuple.get_item(idx).map(|i| i.is_instance_of::<PyVector>()))
+            .all(|i| i.and_then(|i| i).unwrap_or(false))
```
Comment (MichaelScofield) [inline src/script/src/python/pyo3/copr_impl.rs:175]: can we:

```suggestion
            .all(|i| matches!(i, Some(true))).unwrap_or(false))
```

File: src/script/src/python/pyo3/vector_impl.rs (hunk 4 above)
Comment (discord9) [reply src/script/src/python/pyo3/vector_impl.rs:326]: > can this handle `v[:-2]`?
> 
> BTW, for simplicy, I think we need not special handle `1` (like step = 1 or stop = -1), and leave it to rust `Range`. We only need to take care of negative ranges. WDYT

After call `slice.indices(len)` stop is already adjusted to positive indices( So stop==-1 actually means before array's head), but it's reasonable to add related tests, gonna add some

File: src/script/src/python/ffi_types/copr.rs (hunk 5 above)
Comment (discord9) [reply src/script/src/python/ffi_types/copr.rs:361]: Hi there's copilot's suggestiong! didn't think much and pressed tab XD

File: src/script/src/python/ffi_types/pair_tests/sample_testcases.rs (hunk 2 above)
Comment (discord9) [reply src/script/src/python/ffi_types/pair_tests/sample_testcases.rs:83]: added something like this:
```python
df = PyDataFrame.from_sql("select number from numbers limit 5").filter(col("number") > 2)
collected = df.collect()
assert len(collected[0][0]) == 2
```

> can we add a case that combine SQL and dataframe API? like
> 
> ```python
> df = PyDataFrame.from_sql("blabla").filter("blabla")
> ```
> 
> It should be supported now.

File: src/script/src/python/pyo3/vector_impl.rs [hunk 8]
```diff
@@ -299,6 +299,49 @@ impl PyVector {
...
+            let vector = self.as_vector_ref();
+
+            let mut buf = vector
+                .data_type()
+                .create_mutable_vector(indices.slicelength as usize);
+            let v = if indices.slicelength == 0 {
+                buf.to_vector()
+            } else {
+                if indices.step > 0 {
+                    let range = if stop == -1 {
```
Comment (MichaelScofield) [inline src/script/src/python/pyo3/vector_impl.rs:324]: Can we calculate `stop` from `slicelength`? It seems `|stop - start| = slicelength`?

Comment (discord9) [reply src/script/src/python/pyo3/vector_impl.rs:324]: I originally thought Python's slicing length is the final results of slicing, like floor((stop-start)/step), gonna check if this is correct

**Review by waynexia — APPROVED** [review]
Other places look good to me 🚀

File: src/script/src/python/pyo3/vector_impl.rs (hunk 8 above)
Comment (discord9) [reply src/script/src/python/pyo3/vector_impl.rs:324]: > Can we calculate `stop` from `slicelength`? It seems `|stop - start| = slicelength`?

https://docs.python.org/3/c-api/slice.html#c.PySlice_GetIndicesEx
it seems slicelength store length of slice, so unless step==1 it's not equal?

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**